File: zad4/Polynomial.py
class Polynomial:
  lst = []
  
  def __init__(self, lst):
    self.lst = lst.copy()
    Polynomial.trimList(lst)

  def degree(self):
    return len(self.lst) - 1
  
  def __add__(self, pol):
    lst1 = [x for x in self.lst]
    lst2 = [x for x in pol.lst]
    Polynomial.alignLists(lst1, lst2)
    result = [lst1[i] + lst2[i] for i in range(len(lst1))]
    Polynomial.trimList(result)
    return Polynomial(result)
  
  def __neg__(self):
    result = [-x for x in self.lst]
    return Polynomial(result)

  def __sub__(self, number):
    temp = -number
    return self + temp
  
  def __mul__(self, number):
    selfCopy = Polynomial([x for x in self.lst])
    classType = type(selfCopy.lst[0])
    result = Polynomial([classType(0)])
    Polynomial.alignLists(result.lst, selfCopy.lst)

    for i in range(len(number.lst)):
      temp = Polynomial(selfCopy.lst)
      Polynomial.scalarMult(temp.lst, number.lst[i])
      result = result + temp
      Polynomial.shiftList(selfCopy.lst)
    
    Polynomial.trimList(result.lst)
    return result

  def __truediv__(self, number):
    return Polynomial.divAlg(self, number)[0]

  def __mod__(self, number):
    return Polynomial.divAlg(self, number)[1]

  @staticmethod
  def divAlg(left, right):
    zeroElem  = type(left.lst[0])(0)
    if left.degree() == 0 and right.lst[0] == zeroElem:
      raise ArithmeticError

    remainder = Polynomial([x for x in left.lst])
    quotient = Polynomial([zeroElem])
    quotientSub = [x for x in right.lst]
    quotientSub.pop()
    quotientSub = Polynomial(quotientSub)

    while remainder.degree() > right.degree():
      temp = []
      while len(temp) < remainder.degree() + 1:
        temp.append(zeroElem)
      
      temp[remainder.degree() - right.degree()] = remainder.lst[len(remainder.lst) - 1] / right.lst[len(right.lst) - 1]
      quotient = quotient + Polynomial(temp)
      remainder.lst.pop(len(left.lst) - 1)
      Polynomial.trimList(quotient.lst)
      remainder = remainder - (quotientSub * Polynomial(temp))
    return quotient, remainder

  def __str__(self):
    counter = 0
    result = ""
    for x in self.lst:
      result += f"{x}*x^{counter}"
      counter += 1
      if counter < len(self.lst):
        result += " + "
    return result

  @staticmethod
  def trimList(lst : list):
    if lst == []:
      return
    resType = type(lst[0])
    zero = resType(0)
    
    while len(lst) > 1 and lst[-1] == zero:
      lst.pop()
  
  @staticmethod
  def alignLists(lst1 : list, lst2 : list):
    classType = type(lst1[0])
    while len(lst1) < len(lst2):
      lst1.append(classType(0))
    
    while len(lst2) < len(lst1):
      lst2.append(classType(0))
  
  @staticmethod
  def shiftList(lst : list):
    classType = type(lst[0])
    lst.insert(0, classType(0))
  
  @staticmethod
  def scalarMult(lst, scalar):
    for i in range(len(lst)):
      lst[i] *= scalar

File: zad3/Polynomial.java
package zad3;

import java.util.ArrayList;
import java.util.List;

public class Polynomial<T extends Field<T>> {
  private List<T> coefficients = null;

  public Polynomial(List<T> coefficients) {
    this.coefficients = new ArrayList<>(coefficients);
    trimList(this.coefficients);
  }

  public Polynomial<T> sum(Polynomial<T> pol) {
    List<T> coefficientsCopy = new ArrayList<>(this.coefficients);
    List<T> polCopy = new ArrayList<>(pol.coefficients);
    List<T> result = new ArrayList<>();
    alignLists(coefficientsCopy, polCopy);

    for (int i = 0; i < polCopy.size(); i++) {
      result.add(coefficientsCopy.get(i).sum(polCopy.get(i)));
    }
    trimList(result);

    return new Polynomial<T>(result);
  }

  public Polynomial<T> negate() {
    List<T> temp = new ArrayList<>();
    for (var it : this.coefficients) {
      temp.add(it.negate());
    }

    return new Polynomial<>(temp);
  }

  public Polynomial<T> diff(Polynomial<T> pol) {
    var temp = pol.negate();
    return this.sum(temp);
  }

  public Polynomial<T> product(Polynomial<T> pol) {
    Polynomial<T> thisCopy = new Polynomial<>(new ArrayList<>(this.coefficients));
    Polynomial<T> result = new Polynomial<>(new ArrayList<>());
    result.coefficients.add(this.coefficients.get(0).zeroElem());
    alignLists(result.coefficients, thisCopy.coefficients);

    for (int i = 0; i < pol.coefficients.size(); i++) {
      var temp = new Polynomial<T>(thisCopy.coefficients);
      scalarMult(temp.coefficients, pol.coefficients.get(i));
      result = result.sum(temp);
      shift(thisCopy.coefficients);
    }

    trimList(result.coefficients);
    return result;
  }

  public Polynomial<T> div(Polynomial<T> right) throws ArithmeticException{
    return divAlgorithm(this, right).first;
  }

  public Polynomial<T> rem(Polynomial<T> right) throws ArithmeticException{
    return divAlgorithm(this, right).second;
  }

  private Pair divAlgorithm(Polynomial<T> left, Polynomial<T> right) {
    var zeroElem = left.coefficients.get(0).zeroElem();
    if (right.degree() == 0 &&
        right.coefficients.get(0).compareTo(zeroElem) == 0) {
      throw new ArithmeticException("Zero division");
    }

    Polynomial<T> remainder = new Polynomial<>(left.coefficients);
    Polynomial<T> quotient = new Polynomial<>(new ArrayList<>());
    quotient.coefficients.add(zeroElem.zeroElem());
    Polynomial<T> quotientSub = new Polynomial<>(right.coefficients);
    quotientSub.coefficients.remove(quotientSub.coefficients.size() - 1);

    while (remainder.degree() >= right.degree()) {
      List<T> temp = new ArrayList<>();
      
      while (temp.size() < remainder.degree() + 1) {
        temp.add(zeroElem.zeroElem());
      }

      temp.set(remainder.degree() - right.degree(),
              remainder.coefficients.get(remainder.coefficients.size() - 1).division(
                right.coefficients.get(right.coefficients.size() - 1)
              ));
      quotient = quotient.sum(new Polynomial<>(temp));
      remainder.coefficients.remove(remainder.coefficients.size() - 1);
      trimList(quotient.coefficients);
      remainder = remainder.diff(quotientSub.product(new Polynomial<>(temp)));
    }

    return new Pair(quotient, remainder);
  }

  public int degree() {
    return this.coefficients.size() - 1;
  }

  @Override
  public String toString() {
    String result = "";

    int counter = 0;
    for (var it : this.coefficients) {
      result += it.toString() + "*x^" + counter; 
      counter++;
      if (counter < this.coefficients.size()) {result += " + ";}
    }

    return result;
  }

  private void trimList(List<T> t) {
    while (t.size() > 1 && 
           t.get(t.size() - 1).compareTo(t.get(0).zeroElem()) == 0) {
      t.remove(t.size() - 1);
    }
  }

  private void alignLists(List<T> lst1, List<T> lst2) {
    while (lst1.size() < lst2.size()) {
      lst1.add(lst1.get(0).zeroElem());
    }

    while (lst2.size() < lst1.size()) {
      lst2.add(lst2.get(0).zeroElem());
    }
  }

  private void scalarMult(List<T> lst, T scalar) {
    for(int i = 0; i < lst.size(); i++) {
      lst.set(i, lst.get(i).product(scalar));
    }
  }

  private void shift(List<T> lst) {
    lst.add(0, lst.get(0).zeroElem());
  }

  private class Pair {
    public Polynomial<T> first;
    public Polynomial<T> second;
    Pair(Polynomial<T> f, Polynomial<T> s) {
      this.first = f;
      this.second = s;
    }
  }
}
